### ISSUE TITLE
Mongo doesn't trigger any listener

### ISSUE THREAD

**Issue by nicutimofte** [typeorm/typeorm#1527]
**Issue type:**

[x] question
[ ] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[x] `mongodb`
[ ] `mssql`
[x] `mysql` / `mariadb`
[ ] `oracle`
[ ] `postgres`
[ ] `sqlite`
[ ] `sqljs`
[ ] `websql`

**TypeORM version:**

[x] `latest`
[ ] `@next`
[ ] `0.x.x` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**

 I am working with 2 connections: mysql and mongodb. My listeners are triggered only by the mysql operations: insert, update etc. but mongodb is not triggering any listeners. My mongodb version is ```mongodb: ^2.2.33```

```
@EventSubscriber()
export class EntitySubscriber implements EntitySubscriberInterface {

  public beforeInsert(event: InsertEvent<any>) {
    console.log('*before insert', event.entity)
  }

  public beforeUpdate(event: UpdateEvent<any>) {
    console.log('*before update', event.entity)
  }

  public beforeRemove(event: RemoveEvent<any>) {
    console.log('*before remove', event.entity)
  }
}
```

I also have entity listeners inside the entity classes, which does not trigger too.
```

@Entity('componentsContent')
export class ComponentContent extends BaseEntity {

  @ObjectIdColumn()
  public id: ObjectID

  @Column('json')
  public content: any

  @CreateDateColumn()
  public createdAt: Date

  @UpdateDateColumn()
  public updatedAt: Date

  @VersionColumn()
  public version: number

  @AfterUpdate()
  public updated() {
    console.log("updated")
  }

  @AfterInsert()
  public inserted() {
    console.log("inserted")
  }

  @AfterLoad()
  public loaded() {
    console.log('loaded')
  }
}
```

The subscriber class is registered in the ormconfig:
```
subscribers: [
    EntitySubscriber
  ]
```

Do I need any specific version of mongo ? Or I am missing something ?

Thank you!

**Comment by teonapster** [comment]
Any resolution to this issue? What i did to solve it (temporarily) was to create a wrapper class which wraps 
 all mongo actions (create,update,find,delete etc) and manual calls afterLoad, afterRemove etc methods

**Comment by pleerock** [comment]
This is opened for user contributions!

**Comment by xmlking** [comment]
My Workaround: 

project.entity.ts
```ts
@Entity()
export class Project  {
  @ObjectIdColumn() id: ObjectID;
  @Index()
  @Column({nullable: false}) name: string;

  @CreateDateColumn()
  createdAt?: Date;

  @UpdateDateColumn()
  updatedAt?: Date;

  @Exclude()
  @VersionColumn()
  version: number;
}
```
project.repository.ts

```ts
import { Connection, EntityRepository, MongoRepository } from 'typeorm';
import { Project } from './project.entity';
import { Injectable } from '@nestjs/common';
import { DeepPartial } from 'typeorm/common/DeepPartial';
import { SaveOptions } from 'typeorm/repository/SaveOptions';
import { UpdateResult } from 'typeorm/query-builder/result/UpdateResult';

@Injectable()
@EntityRepository(Project)
export class ProjectRepository extends MongoRepository<Project> {
  save<T extends DeepPartial<Project>>(entity: T, options?: SaveOptions): Promise<T> {
    const createdCol = this.metadata.createDateColumn ? this.metadata.createDateColumn.propertyPath : null;
    const versionCol = this.metadata.versionColumn ? this.metadata.versionColumn.propertyPath : null;
    if (createdCol) {
      (entity as any)[createdCol] = new Date();
    }
    if (versionCol) {
      (entity as any)[versionCol] = 1;
    }
    return super.save(entity, options);
  }

  async update(
    criteria: string,
    partialEntity: DeepPartial<Project>,
    options?: SaveOptions,
  ): Promise<UpdateResult> {
    const updatedCol = this.metadata.updateDateColumn ? this.metadata.updateDateColumn.propertyPath : null;
    const versionCol = this.metadata.versionColumn ? this.metadata.versionColumn.propertyPath : null;
    if (updatedCol) {
      (partialEntity as any)[updatedCol] = new Date();
    }
    if (versionCol) {
      const found = await super.findOne(criteria);
      if (found) {
        (partialEntity as any)[versionCol] = found[versionCol] + 1;
      }
    }
    return super.update(criteria, partialEntity, options);
  }
}

export const ProjectRepositoryProvider = {
  provide: 'ProjectRepository',
  useFactory: (connection: Connection) => connection.getCustomRepository(ProjectRepository),
  inject: [Connection],
};
```

project.module.ts
```ts
import { Module } from '@nestjs/common';
import { ProjectService } from './project.service';
import { ProjectController } from './project.controller';
import { TypeOrmModule } from '@nestjs/typeorm';
import { Project } from './project.entity';
import { ConfigModule } from '../config';
import { ProjectRepository, ProjectRepositoryProvider } from './project.repository';
import { getManager, getCustomRepository } from 'typeorm';

@Module({
  imports: [ConfigModule, TypeOrmModule.forFeature([Project])],
  providers: [
    ProjectService,
    ProjectRepositoryProvider,
  ],
  controllers: [ProjectController],
})
export class ProjectModule {}
```


project.service.ts
```ts
import { Inject, Injectable, Logger, OnModuleDestroy, OnModuleInit } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { In, MongoRepository, SelectQueryBuilder } from 'typeorm';
import { Project } from './project.entity';
import { ProjectDto } from './project.dto';
import { ProjectRepository } from './project.repository';

@Injectable()
export class ProjectService implements OnModuleInit, OnModuleDestroy {
  private readonly logger = new Logger(ProjectService.name);
  constructor(
    private readonly projectRepository: ProjectRepository,
  ) {
  }

  async findAll(): Promise<Project[]> {
    return await this.projectRepository.find();
  }
  async findOne(id: string): Promise<Project> {
    return await this.projectRepository.findOne(id);
  }
  async create(project: ProjectDto): Promise<Project> {
    const obj = this.projectRepository.create(project);
    return await this.projectRepository.save(obj);
  }
  async replace(id: string, project: ProjectDto): Promise<any> {
    return this.projectRepository.update(id, project);
  }
  async update(id: string, project: Partial<ProjectDto>): Promise<any> {
    return this.projectRepository.update(id, project);
  }

}
```

**Comment by zuijiaoluo** [comment]
I just used it, I don't know how the underlying layer is done. Can I have some hints, I can contribute code, because this is a feature that many people need. And he is very important to me.

**Comment by rustamwin** [comment]
:+1:

**Comment by zuijiaoluo** [comment]
This is a new problem. When I save, beforeInsert, beforeUpdate, and afterUpdate are triggered at the same time. I think beforeUpdate and afterUpdate should not be triggered in save

**Comment by rustamwin** [comment]
> if i use the update or updateOne method

use `save` instead of `update`
> Using save method is not working because the manager try to insert instead of updating. Even if the id is present.

Hmm... should work

**Comment by Rmannn** [comment]
I have to mention that my primary key is an ObjectIDColumn({name: '_id'}) id: ObjectID

**Comment by Rmannn** [comment]
I have an other `ObjectIdColumn({name: 'zoneId'}) zoneId: ObjectID;`
It seems the framework is trying to insert the entity instead to update it.
If i remove this property, every thing works. Do you know why ?

**Comment by Rmannn** [comment]
Ok the problem is inside the ObjectIdColumn,

The options.primary is force to true. So if an other field is declared using this decorator, thhe persistent layer will fail.
The first thing to do is to create a custom ObjectIdColumn allowing us to specify the options primary.
```ts
import { ColumnOptions, getMetadataArgsStorage } from 'typeorm';

export function ObjectIdSecondaryColumn(options?: ColumnOptions): any {
    return (object: any, propertyName: string) => {
        if (!options) {
            options = {};
        }
        if (!options.name) {
            options.name = propertyName;
        }
        getMetadataArgsStorage().columns.push({
            target: object.constructor,
            propertyName,
            mode: 'objectId',
            options
        });
    };
}
```
The second pb is from the class-transformer that not works well with object id
A pull request is waiting to fix that
https://github.com/typestack/class-transformer/issues/53

**Comment by Rmannn** [comment]
Did you try adding an other property typed as ObjectIdColumn and optional ?
If you try to update this kind of entity, the save method will try to create the object instead updating it. Resulting in a Duplicate Error (Cause the object already exists)
Because the ObjectIdColumn works only for primary column and this should be documented.

However, Thank you for your time and for this awesome library.
We decide to go further directly with the mongo db driver.

**Comment by rustamwin** [comment]
@Rmannn I think this problem fixed #3291

**Comment by Rmannn** [comment]
Thx, I will keep it for my others projects !

**Comment by vlapo** [comment]
Should be fixed in release `0.2.12`
